source/generate.py

```diff
@@ -92,11 +66,11 @@ def generate_one_sequence(
         verbose=True,
         expected_length=8,
     ):
-        """generate a sequence based on
-        - the input_prompt and/or inst and density parameters
-        - the "final prompt" is converted into input_prompt_ids
+        """generate a sequence
+        - input_prompt combined with inst and density parameters -> input_prompt
+        - input_prompt is converted into input_prompt_ids
         - input_prompt_ids are passed to generate_sequence_of_token_ids for generation
-        - the generated toekn_ids are then converted to text"""
+        - the generated token_ids are then converted to text"""
 
         if inst is not None:
```

Comment:
Could this be simplified by assumming a density must always be given for any instrument?
You zip them both in an iterator earlier, which means inst list and density list must be the same size

Reply:
Actually not really:
We want to be able to prompt:
- "PIECE_START " -> Instrument does not have to be prompted
- "PIECE_START INST=X" -> Instrument alone can be prompted
- "PIECE_START INST=X DENSITY=Y" -> Instrument and density can be prompted together

Right now, if an instrument is not defined then density cannot be defined.("PIECE_START _____ DENSITY=Y"). This comes from the fact that the prompt needs to be in the defined order PIECE_START INST=X DENSITY=Y ... we could try to prompt in a different order I am not sure how it would work... pretty sure it would not behave very well

One way to solve this could be to stochastically choose an instrument based on their occurence in the trained dataset for example, then add it to the prompt, and then add the chosen density...